portset/portset.go

```diff
@@ -0,0 +1,225 @@
+package portset
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os/user"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aporeto-inc/trireme-lib/cache"
+	"github.com/bvandewalle/go-ipset/ipset"
+	"go.uber.org/zap"
+)
+
+const (
+	procNetTCPFile                    = "/proc/net/tcp"
+	portSetUpdateIntervalMilliseconds = 1000
+	portEntryTimeout                  = portSetUpdateIntervalMilliseconds * 3
+	uidFieldOffset                    = 7
+	procHeaderLineNum                 = 0
+	portOffset                        = 1
+	ipPortOffset                      = 1
+	sockStateOffset                   = 3
+	sockListeningState                = "0A"
+	hexFormat                         = 16
+	integerSize                       = 64
+	decFormat                         = 10
+)
+
+// portSetInstance : This type contains look up tables
+// to help update the ipset portsets.
+type portSetInstance struct {
+	userPortSet *cache.Cache
+
+	userPortMap *cache.Cache
+}
+
+// deletes the port entry in the portset when the key uid:port
+// expires
+func expirer(c cache.DataStore, id interface{}, item interface{}) {
+
+	userPort := strings.Split(id.(string), ":")
+	portSetObject := item.(*portSetInstance)
+	user := userPort[0]
+	port := userPort[1]
+
+	if err := portSetObject.deletePortSet(user, port); err != nil {
+		zap.L().Warn("Failed to delete port from set", zap.Error(err))
+	}
+
+}
+
+// New creates a portset interface
+func New() PortSet {
+
+	p := &portSetInstance{
+		userPortSet: cache.NewCache("userPortSet"),
+		userPortMap: cache.NewCacheWithExpirationNotifier("userPortMap", portEntryTimeout, expirer),
+	}
+
+	go startPortSetTask(p)
+
+	return p
+}
+
+func getUserName(uid string) (s string, err error) {
+
+	if u, err := user.LookupId(uid); err == nil {
+		return u.Username, nil
+	}
+	return "", err
+}
+
+// AddPortToUser adds/updates userPortMap cache. returns
+// true if user key is already present
+func (p *portSetInstance) AddPortToUser(userName string, port string) (ok bool, err error) {
+
+	ok = false
+	key := userName + ":" + port
+
+	if _, err := p.userPortMap.Get(key); err == nil {
+		ok = true
+	} else {
+		p.userPortMap.Add(key, p)
+	}
+
+	// program the ipset
+	if err = p.addPortSet(userName, port); err != nil {
+		return false, err
+	}
+
+	return ok, nil
+}
+
+// AddUserPortSet : This adds/updates userPortSet cache. This cache
+// maps user to the portset associated with its PU. This is called during
+// creation of PU/on reception of application SYN-ACK packet.
+func (p *portSetInstance) AddUserPortSet(userName string, portset string) (err error) {
+
+	p.userPortSet.AddOrUpdate(userName, portset)
+	return nil
+
+}
+
+// GetUserPortSet returns the portset associated with user.
+func (p *portSetInstance) getUserPortSet(userName string) (port string, err error) {
+
+	portSetName, err := p.userPortSet.Get(userName)
```

Comment:
1) Merge if into same line.

2) if typecast of portSetName to string fails ? we return nil error and nil port